File: Daily/2020Nov/convert-binary-number-in-a-linked-list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function getDecimalValue(head: ListNode | null): number {
  const binary = returnBinary(head, "");
  return parseInt(binary, 2);
}

const returnBinary = (head: ListNode | null, res: string): string => {
  if (head === null) {
    return res;
  }
  return returnBinary(head.next, `${res}${head.val}`);
};

File: src/bin/convert-binary-number-in-a-linked-list.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

fn node_checker(head: Option<Box<ListNode>>, mut str: String) -> String {
    match head {
        Some(x) => {
            str.push(std::char::from_digit(x.val as u32, 10).unwrap());
            node_checker(x.next, str)
        }
        None => str,
    }
}

struct Solution {}

impl Solution {
    pub fn get_decimal_value(head: Option<Box<ListNode>>) -> i32 {
        use std::i32;
        let res = node_checker(head, String::new());
        i32::from_str_radix(&res[..], 2).unwrap()
    }
}

fn main() {}
